### DISCUSSION TITLE
criticalUpdate does not force an update if there is a newer "non" criticalUpdate...

### DISCUSSION THREAD

**Discussion by BobBrownConsulting** [sparkle-project/Sparkle#2634] [Q&A] [answered]
As part of my testing, I created 3 builds. The first (build 11) was the first update. The second (build 12) was a critical update that should force the user to update. The third (build 14) was an update (not forced).

When I ran build 11, I got the standard update UI to update to 14 with the option to skip.

**My expectation was that I might be able to skip build 14, BUT I would be required to update to build 12 (at a minimum).**

It could be that I did not generate the appcast file exactly right.

For critical updates, I use:

generate_appcast /sparkle/ --critical-update-version $BUILD_NUMBER

For standard updates, I use:

generate_appcast /sparkle/ 

It might make sense to use --maximum-versions or similar options to get the behavior I desire.

The man page of generate_appcast doesn't quite indicate how I should use the settings for this workflow.

Any guidance or experience is much appreciated.

[appcast.xml.zip](https://github.com/user-attachments/files/17120032/appcast.xml.zip)

**Comment by BobBrownConsulting** [comment]
It seems that I need to remove version before my critical update to get the behavior I'm desiring.

generate_appcast /sparkle --critical-update-version 1.0.0.12 --maximum-versions 2

or even

generate_appcast /sparkle --critical-update-version 1.0.0.12 --maximum-versions 1

**Comment by zorgiepoo** [comment, marked as answer]
`maximum-versions` is irrelevant here. Once Sparkle chooses the best update item from your appcast, the other ones are not referenced again. So the appcast item that is chosen needs to have all the sufficient information on the critical update.

Your builds are:

```
1.0.0.14
1.0.0.12
1.0.0.11
```

..according to the appcast you uploaded (from the `sparkle:version`). There are no builds that are 11, 12, 14.

The appcast you uploaded indicates you passed `--critical-update-version 12` for update `1.0.0.12` which would be wrong.

`--critical-update-version 1.0.0.12` should make update `1.0.0.14` appear critical if the user has a build < 1.0.0.12. It does not offer the user update to `1.0.0.12` first (this is not desirable). Note your chosen update (`1.0.0.14` in this case) update needs the `<sparkle:criticalUpdate sparkle:version="1.0.0.12"/>`


--

> It seems that I need to remove version before my critical update to get the behavior I'm desiring.

Re-running generate_appcast again with --critical-update-version probably doesn't override the appcast item you already created from a previous run. So you might have found out you had to either modify the appcast manually or remove the item and run it again..
